src/bridge/cockpitpackages.c

```diff
@@ -400,20 +418,14 @@ read_package_manifest (const gchar *directory,
     }
   else
     {
-      override = read_json_file (directory, "override.json", &error);
-      if (error)
-        {
-          if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-            g_debug ("%s: no override found", package);
-          else
-            g_message ("%s: couldn't read override.json: %s", package, error->message);
-          g_clear_error (&error);
-        }
-      if (override)
-        {
-          cockpit_json_patch (manifest, override);
-          json_object_unref (override);
-        }
+      /* possible override locations, in ascending priority */
+      apply_override (manifest, directory, "override.json");
+
+      g_autofree gchar *package_override_name = g_strconcat (package, ".override.json", NULL);
+      apply_override (manifest, PACKAGE_SYSCONF_DIR "/cockpit", package_override_name);
+
+      g_autofree gchar *cockpit_user_config = g_build_filename (g_get_user_config_dir (), "cockpit", NULL);
```

Comment:
I find the multiple layers of path pasting here to be a bit distasteful.  Let's modify `json_read_file()` to rather take a full pathname and perform all path computations here, or modify it to take a `const gchar **` and use compound literals (which would let us keep everything here `const`).  I think `...` will be too much effort, because there's no variant of the glib path building (or even `g_strjoin()`) which takes a va_args.

Reply:
There is [g_build_filename_valist()](https://docs.gtk.org/glib/func.build_filename_valist.html). This has more runtime penalty and increases the amount of code, and IMHO this is not worth it for this case. Likewise, [g_build_filenamev()](https://docs.gtk.org/glib/func.build_filenamev.html) is overkill here -- single-time construction and iteration of whole arrays.

So I'll pull the g_build_filename() out of read_json_file(). It will make the read_package_manifest() a little bigger, but still readable and avoid unnecessary effort or unwieldy syntax.

Reply:
I agree.  Path-pasting in a post-`g_autofree` world is pretty reasonable.